### DISCUSSION TITLE
0.15.0

### DISCUSSION THREAD

**Discussion by MohamedBassem** [karakeep-app/karakeep#276] [Announcements]
Welcome to the `0.15.0` release of Hoarder! It's been a while since the previous release, but this release is packed with a lot of new quality of life improvements and features. This releases wouldn't have been possible without the tons of contributions from @kamtschatka, @mdsaban & @AhmadMuj. I also want to welcome our first time contributors @kou029w & @devome. If you're interested in helping us develop hoarder, you can join our #development channel on [hoarder's discord](https://discord.gg/NrgeYywsFh).

# New Features

* Bulk actions are now a thing thanks to a contribution from @mdsaban. You can now bulk favourite, archive or delete bookmarks. In the next release, you'll be able to bulk add/remove bookmarks from lists.
* Full page archival is now here (powered by [monolith](https://github.com/Y2Z/monolith)). This will download everything that's in the page so that you can have an offline copy of that page even if that page disappears from the internet afterwards. The archives are typically quite large, so this feature is disabled by default. If you're a true hoarder, you can enable it with `CRAWLER_FULL_PAGE_ARCHIVE=true`.
* When you hoard a URL that points to a pdf or an image, Hoarder will now treat them as pdfs/images instead of links while keeping a reference to their original source. This will also allow automatic tagging to work correctly on those links. (thanks @kamtschatka!)
* You can now import your chrome bookmarks (or any Netscape compatible bookmark file) by exporting them from chrome, then drag and drop the exported file to hoarder. All links in the export will be imported. (thanks @kamtschatka!)
* For those hoarding code snippets, @kamtschatka added syntax highlighting to code blocks and a way to quickly copy the block.
* Hoarder is now available on arch linux's user repo (AUR) and is maintained by @devome. A new wiki page with [installation instructions](https://docs.hoarder.app/Installation/archlinux) is now available.

# UX Improvements

* @mdsaban revamped how both the user and admin settings pages look like.
* In the list layout, the editor will expand as you type longer notes thanks to @AhmadMuj.
* You can now paste images directly in the editor card and they'll get automatically uploaded thanks to @AhmadMuj.

# Fixes

* The layout selector is now more clearly shown above the bookmark grid thanks to @mdsaban.
* Tags that were added manually were not correctly propagating to the search index. This is now fixed by @kamtschatka. You might want to trigger a reindex from the admin page if you've attached some manual tags.
* @kamtschatka fixed a long standing bug that prevented firefox users from using the browser extension without a valid HTTPS certificate.
* The editor card previously allowed you to save empty notes, this has now been fixed by @mdsaban.
* Increased list name max size to 40 chars up from 20.
* We previously had an upper limit on note sizes that we removed a while ago. However, we forgot to remove it when editing a note. @AhmadMuj fixed that.
* The CLI was previously printing invalid JSONs. Now there's a proper JSON mode thanks to @kamtschatka.
* For CJK environments, there's now a sans-serif fallback font thanks to @kou029w.
* Automatically inferred tags could have contained an extra space in the beginning of the tag, @kamtschatka fixed that.

# Screenshots

## Bulk Editing in Action

https://github.com/hoarder-app/hoarder/assets/45597394/d8625dfe-dcb6-4b1f-91a6-8dbe289cebdb

## Revamped Admin Page

![image](https://github.com/hoarder-app/hoarder/assets/2418637/cb375b37-c88e-453f-ab9b-55c2b309fa95)

## Revamped Settings Page

![image](https://github.com/hoarder-app/hoarder/assets/2418637/0aec69b0-89c7-4fc0-8003-a94b5cbb2793)

## Full Page Archives

<img width="1159" alt="Screenshot 2024-07-06 at 11 47 31 AM" src="https://github.com/hoarder-app/hoarder/assets/2418637/ba0c05d8-5194-4e2a-8c7e-bf02669252f1">

## PDF/Image URL recognition

https://github.com/hoarder-app/hoarder/assets/2418637/fa9f2323-f0f3-4ce3-9818-e4e65e32fa54

# Upgrading

To upgrade:
* If you're using HOARDER_VERSION=release, run `docker compose pull && docker compose up -d`.
* If you're pinning it to a specific version, upgrade the version and then run `docker compose pull && docker compose up -d`. 

# All Commits

- fix(web): Add sans-serif fallback for Inter font in CJK environments (#272) by @kou029w  in #72f03b2
- fix: Renaming full page archive asset file doesn't work inside docker. Fixes #273 by @MohamedBassem in #5e9b04a
- feat: Add bulk edit option for bookmarks. Fixes #84 (#259) by @mdsaban in #d193d9b
- fix: fixed an issue that would half the textarea in grid layout after submitting a bookmark (#269) by @kamtschatka in #bf92fa3
- refactor: fixed typos in BookmarkTypes (#268) by @kamtschatka in #4e70fe3
- refactor: drag and drop improvements (#264) by @kamtschatka in #9cd6170
- fix: Increase list name size to max 40 chars up from 20. Fixes #254 by @MohamedBassem in #9883c6b
- fix: Fix duplicate bookmarks in links with assets by @MohamedBassem in #3d495cd
- refactor: added the bookmark type to the database (#256) by @kamtschatka in #e648646
- refactor: remove redundant code from crawler worker and refactor handling of asset types (#253) by @kamtschatka in #ccbff18
- ui: refactor admin settings page (#249) by @mdsaban  in #e107f8b
- ui: Changes for user settings page (#251) by @mdsaban in #a637130
- feature: Automatically transfer image urls into bookmared assets. Fixes #246 by @MohamedBassem in #1071095
- ui(web): move layout selector to main screen (#252) by @mdsaban in #8f0e9b1
- refactor: extract assets into their own database table. #215 (#220) by @kamtschatka in #9ce6958
- fix(landing): image width on landing page (#250) by @mdsaban in #0f54a18
- feature: add support for PDF links. Fixes #28 (#216) by @kamtschatka in #be1b7f7
- fix(web): Fix save action on empty card. Fixes #225 (#243) by @mdsaban in #ccfff6b
- ui: Improve landing page for homepage (#238) by @mdsaban in #16f21bf
- feature(web): Enhance the bookmark textarea in the list layout (#247) by @AhmadMuj in #ac05ecf
- readme: broken readme links (#233) by @mdsaban in #348bd80
- fix: Remove the bookmark update text limit (#244) by @AhmadMuj in #671fc32
- fix(web): text paste bug introduced with image paste (#232) by @mdsaban in #23faf46
- docs: add Arch Linux installation docs (#223) by @devome in #dc19313
- feature(web): support pasting image in the Editor Card (#228) by @AhmadMuj in #6823caf
- release(cli): Bump CLI version to 0.13.5 by @MohamedBassem in #219329b
- release(cli): Bump CLI version to 0.13.4 by @MohamedBassem in #5f263a3
- fix(cli): Bookmark list output is not a valid JSON. Fixes #150 (#181) by @kamtschatka in #cde9726
- refactor: Extract the bookmark polling logic into a separate shared component by @MohamedBassem in #6928800
- fix: Trigger search re-index on bookmark tag manual updates. Fixes #208 (#210) by @kamtschatka in #546139e
- feature(web): Add syntax highlighting to code blocks and a quick copy button. Fixes #195 (#197) by @kamtschatka in #be1bb38
- fix(cli): Fix CLI version in help message. Fixes #139 by @MohamedBassem in #f7a7753
- feature(extension): Add a context menu to open your hoarder saves. Fixes #188 (#206) by @kamtschatka in #2d0af0b
- fix(web): Changed the toast to contain a more detailed error message in case the bookmark is already part of the list. Fixes #179 (#213) by @kamtschatka in #049e79d
- fix(workers): AI infered tags can contain " " at the beginning. Fixes #184 (#194) by @kamtschatka in #be8b91f
- fix(crawler): Only update the database if full page archival is enabled by @MohamedBassem in #9d89f98
- docs: Document the new CRAWLER_FULL_PAGE_ARCHIVE flag by @MohamedBassem in #9198c1b
- feature: Full page archival with monolith. Fixes #132 by @MohamedBassem in #dedc5fb
- feature: Allow import Netscape HTML format (#163) by @kamtschatka in #033e8a2
- build: Let the upload-artificat action do the compression by @MohamedBassem in #bb431be
- release(extension): Bump extension version to 1.2.1 by @MohamedBassem in #8688dd4
- build: Add a workflow to build the extension by @MohamedBassem in #c9b1307
- fix(extension): Drop the https restriction in firefox. Fixes #23 by @kamtschatka in #875cc35
- fix(cli): Switched to cursorV2 to make use of the fix for #140 and prevent infinite loops (#176) by @kamtschatka in #6dc67cf

<hr /><em>This discussion was created from the release <a href='https://github.com/hoarder-app/hoarder/releases/tag/v0.15.0'>0.15.0</a>.</em>

**Comment by akshara-tg** [comment]
Thank you for another release with lot of enhancements. 
My. .env file did not had "CRAWLER_FULL_PAGE_ARCHIVE=true" earlier. I have upgraded to the latest release and added "CRAWLER_FULL_PAGE_ARCHIVE=true" in my .env file.
I have around 1500 bookmarks. To prevent myself from 'link ROT' issues, which option should I use? "Recrawl All Links" Or "Reindex All Bookmarks" ?

Once again, thanks for all the hard work you guys do.

**Reply by kamtschatka** [reply]
Reindex will only update Meilisearch, what you want is a full recrawl. 
Careful: I think that will also run tag inference again, so you might get additional new tags. You might want to disable that

**Comment by akshara-tg** [comment]
@kamtschatka Thanks for the swift response. Are you recommending I should use "Recrawl All Links (Without Inference)"? to avoid regenerating the tags using AI?
I curious to know whether "Recrawl All Links" will replace the existing tags with newly generated tags from AI inference OR will that append to the list of existing tags.

**Reply by kamtschatka** [reply]
ah, i did not remember that there is already such an option, Yes, you might want to use without inference then. Inference does not touch manually created tags, but the code seems to delete  tags that did not come up again in the new inference, so it seems like it will replace existing tags with the newly infered ones.

**Reply by MohamedBassem** [reply]
@akshara-tg As @kamtschatka said, "recrawl without inference" is what you're looking for. If you're going to do a full page archive for 1500 links, you'll probably be looking at 10s of GBs of extra storage, so make sure you have enough space for that. I also recommend giving it a shot on a bunch of links first to check if that's what you're looking for before committing to doing it for 1500 links.

**Reply by akshara-tg** [reply]
@MohamedBassem Sure, I will try for few before recrawl the entire database. Appreciate that you took time reconfirm the approach. The option "Refresh" available for individual bookmark, would that re-fetch the entire webpage? 

Thank you.

**Comment by undaunt** [comment]
Thanks for this, I've now fully replaced Shiori and Memos with the full archive feature.

**Comment by drgribb** [comment]
Thank you for Hoarder! It works fine and brings on a smile when I use it and realize that its controlled be me on MY server. 

Its nice to see that its possible to manipulate entries in bulk. 

A request: Is is possible to enable bulk edit when entries are recalled by tag?
